src/App.tsx

```diff
@@ -1,61 +1,121 @@
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { Todo, User } from './types';
+import { useState } from 'react';
+import { TodoList } from './components/TodoList';
+
+function getUserById(userId: number): User | null {
+  return usersFromServer.find(user => user.id === userId) || null;
+}
+
+const initialTodos: Todo[] = todosFromServer.map(todo => ({
+  ...todo,
+  user: getUserById(todo.userId),
+}));
 
 export const App = () => {
+  const [todos, setTodos] = useState<Todo[]>(initialTodos);
+
+  const [title, setTitle] = useState<string>('');
+  const [hasTitleError, setHasTitleError] = useState(false);
+
+  const [userId, setUserId] = useState<number>(0);
+  const [hasUserIdError, setHasUserIdError] = useState(false);
+
+  const resetAll = () => {
+    setTitle('');
+    setUserId(0);
+    setHasTitleError(false);
+    setHasUserIdError(false);
+  };
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setHasTitleError(false);
+  };
+
+  const handleUserIdChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setUserId(+event.target.value);
+    setHasUserIdError(false);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    setHasTitleError(!title);
+    setHasUserIdError(!userId);
+
+    if (!title || !userId) {
+      return;
+    }
+
+    if (title && userId) {
+      const maxId: number = initialTodos.reduce((m, c) =>
+        m.id > c.id ? m : c,
+      ).id;
+
+      const newTodo: Todo = {
+        id: maxId + 1,
+        title,
+        completed: false,
+        userId: +userId,
+        user: usersFromServer.find(user => +userId === user.id) || null,
+      };
+
+      setTodos(prevTodos => [...prevTodos, newTodo]);
+      resetAll();
+    }
+  };
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/todos" method="POST">
+      <form action="/api/todos" method="POST" onSubmit={handleSubmit}>
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label htmlFor="post-title">Title: </label>
+          <input
+            type="text"
+            data-cy="titleInput"
+            id="post-title"
+            value={title}
+            onChange={handleTitleChange}
+            placeholder="Enter a title"
+          />
+          {hasTitleError && <span className="error">Please enter a title</span>}
```

Comment:
error messages can also be output in variables
```
const EMPTY_TITLE_ERROR = 'Please enter a title';
const EMPTY_USER_ERROR = 'Please choose a user';
```

```suggestion
          {hasTitleError && <span className="error">{EMPTY_TITLE_ERROR}</span>}
```